File: apps/android/core/pairing/src/main/java/com/localagentbridge/android/core/pairing/InitialPairingProof.kt
package com.localagentbridge.android.core.pairing

import java.io.ByteArrayInputStream
import java.math.BigInteger
import java.security.AlgorithmParameters
import java.security.KeyFactory
import java.security.MessageDigest
import java.security.Signature
import java.security.interfaces.ECPublicKey
import java.security.spec.ECGenParameterSpec
import java.security.spec.ECParameterSpec
import java.security.spec.X509EncodedKeySpec
import java.util.Base64

const val INITIAL_PAIRING_PROOF_SCHEME = "p256-sha256-der-v1"

data class InitialPairingClientRequest(
    val scheme: String,
    val protocolVersion: Int,
    val requestId: String,
    val pairingNonce: String,
    val pairingCode: String,
    val runtimeDeviceId: String,
    val runtimePublicKey: String,
    val runtimeKeyFingerprint: String,
    val clientDeviceId: String,
    val clientDeviceName: String,
    val clientPublicKey: String,
    val clientKeyFingerprint: String,
    val transportBinding: String?,
) {
    fun transcript(): ByteArray = InitialPairingProof.clientRequestTranscript(this)

    fun digest(): String = InitialPairingProof.sha256Hex(transcript())
}

data class InitialPairingAcceptedResult(
    val scheme: String,
    val protocolVersion: Int,
    val requestId: String,
    val pairingRequestDigest: String,
    val accepted: Boolean,
    val runtimeDeviceId: String,
    val runtimePublicKey: String,
    val runtimeKeyFingerprint: String,
    val trustedDeviceId: String,
    val message: String,
    val transportBinding: String?,
) {
    fun transcript(): ByteArray = InitialPairingProof.acceptedResultTranscript(this)

    fun digest(): String = InitialPairingProof.sha256Hex(transcript())
}

object InitialPairingProof {
    fun clientRequestTranscript(request: InitialPairingClientRequest): ByteArray {
        require(request.scheme == INITIAL_PAIRING_PROOF_SCHEME) { "unsupported proof scheme" }
        require(request.protocolVersion == 1) { "unsupported protocol version" }
        val binding = canonicalTransportBinding(request.transportBinding)
        requireCanonicalPublicKey(request.runtimePublicKey, request.runtimeKeyFingerprint)
        requireCanonicalPublicKey(request.clientPublicKey, request.clientKeyFingerprint)
        return transcript(
            CLIENT_CONTEXT,
            "scheme" to request.scheme,
            "protocol_version" to request.protocolVersion.toString(),
            "request_id" to request.requestId,
            "pairing_nonce" to request.pairingNonce,
            "pairing_code" to request.pairingCode,
            "runtime_device_id" to request.runtimeDeviceId,
            "runtime_public_key" to request.runtimePublicKey,
            "runtime_key_fingerprint" to request.runtimeKeyFingerprint,
            "client_device_id" to request.clientDeviceId,
            "client_device_name" to request.clientDeviceName,
            "client_public_key" to request.clientPublicKey,
            "client_key_fingerprint" to request.clientKeyFingerprint,
            "transport_binding" to binding,
        )
    }

    fun acceptedResultTranscript(result: InitialPairingAcceptedResult): ByteArray {
        require(result.scheme == INITIAL_PAIRING_PROOF_SCHEME) { "unsupported proof scheme" }
        require(result.protocolVersion == 1) { "unsupported protocol version" }
        require(result.accepted) { "only accepted pairing results are signed" }
        require(isLowercaseHex(result.pairingRequestDigest, 64)) { "invalid pairing request digest" }
        val binding = canonicalTransportBinding(result.transportBinding)
        requireCanonicalPublicKey(result.runtimePublicKey, result.runtimeKeyFingerprint)
        return transcript(
            RESULT_CONTEXT,
            "scheme" to result.scheme,
            "protocol_version" to result.protocolVersion.toString(),
            "request_id" to result.requestId,
            "pairing_request_digest" to result.pairingRequestDigest,
            "accepted" to "true",
            "runtime_device_id" to result.runtimeDeviceId,
            "runtime_public_key" to result.runtimePublicKey,
            "runtime_key_fingerprint" to result.runtimeKeyFingerprint,
            "trusted_device_id" to result.trustedDeviceId,
            "message" to result.message,
            "transport_binding" to binding,
        )
    }

    fun verifyAcceptedResult(
        result: InitialPairingAcceptedResult,
        signatureBase64: String,
        expectedRequestId: String,
        expectedPairingRequestDigest: String,
        expectedTrustedDeviceId: String,
        expectedTransportBinding: String?,
    ): Boolean = runCatching {
        val actualBinding = canonicalTransportBinding(result.transportBinding)
        val expectedBinding = canonicalTransportBinding(expectedTransportBinding)
        if (result.requestId != expectedRequestId ||
            result.pairingRequestDigest != expectedPairingRequestDigest ||
            result.trustedDeviceId != expectedTrustedDeviceId ||
            actualBinding != expectedBinding
        ) return false

        val publicKey = requireCanonicalPublicKey(result.runtimePublicKey, result.runtimeKeyFingerprint)
        val signatureBytes = decodeCanonicalBase64(signatureBase64)
        requireCanonicalEcdsaDer(signatureBytes)
        Signature.getInstance("SHA256withECDSA").run {
            initVerify(publicKey)
            update(result.transcript())
            verify(signatureBytes)
        }
    }.getOrDefault(false)

    internal fun requireCanonicalPublicKey(publicKeyBase64: String, fingerprint: String): ECPublicKey {
        require(isLowercaseHex(fingerprint, 64)) { "fingerprint must be 64 lowercase hex characters" }
        val encoded = decodeCanonicalBase64(publicKeyBase64)
        val publicKey = KeyFactory.getInstance("EC")
            .generatePublic(X509EncodedKeySpec(encoded)) as? ECPublicKey
            ?: throw IllegalArgumentException("public key must be an EC key")
        require(publicKey.encoded.contentEquals(encoded)) { "public key encoding is not canonical" }
        require(publicKey.params.matchesP256()) { "public key must use P-256" }
        require(sha256Hex(encoded) == fingerprint) { "public key fingerprint mismatch" }
        return publicKey
    }

    internal fun sha256Hex(bytes: ByteArray): String = MessageDigest.getInstance("SHA-256")
        .digest(bytes)
        .joinToString("") { "%02x".format(it) }

    private fun transcript(context: String, vararg fields: Pair<String, String>): ByteArray {
        val lines = ArrayList<String>(1 + fields.size * 3)
        lines += context
        fields.forEach { (name, value) ->
            lines += name
            lines += value.toByteArray(Charsets.UTF_8).size.toString()
            lines += value
        }
        return lines.joinToString("\n").toByteArray(Charsets.UTF_8)
    }

    private fun canonicalTransportBinding(value: String?): String {
        if (value == null || value == "none") return "none"
        require(isLowercaseHex(value, 64)) { "transport binding must be 64 lowercase hex characters or none" }
        return value
    }

    private fun decodeCanonicalBase64(value: String): ByteArray {
        val decoded = Base64.getDecoder().decode(value)
        require(Base64.getEncoder().encodeToString(decoded) == value) { "base64 is not canonical" }
        return decoded
    }

    private fun requireCanonicalEcdsaDer(der: ByteArray) {
        val input = ByteArrayInputStream(der)
        require(input.read() == 0x30) { "signature must be a DER sequence" }
        val sequenceLength = readDerLength(input)
        require(sequenceLength == input.available()) { "invalid DER sequence length" }
        readCanonicalDerInteger(input)
        readCanonicalDerInteger(input)
        require(input.available() == 0) { "trailing DER data" }
    }

    private fun readCanonicalDerInteger(input: ByteArrayInputStream) {
        require(input.read() == 0x02) { "signature component must be a DER integer" }
        val length = readDerLength(input)
        require(length in 1..33 && length <= input.available()) { "invalid DER integer length" }
        val bytes = ByteArray(length)
        require(input.read(bytes) == length)
        require(bytes[0].toInt() and 0x80 == 0) { "DER integer must be positive" }
        require(!(length > 1 && bytes[0] == 0.toByte() && bytes[1].toInt() and 0x80 == 0)) {
            "DER integer is not minimally encoded"
        }
        val value = BigInteger(1, bytes)
        require(value.signum() > 0 && value < P256_ORDER) { "ECDSA component is out of range" }
    }

    private fun readDerLength(input: ByteArrayInputStream): Int {
        val first = input.read()
        require(first >= 0 && first and 0x80 == 0) { "DER length must use canonical short form" }
        return first
    }

    private fun ECParameterSpec.matchesP256(): Boolean {
        val expected = AlgorithmParameters.getInstance("EC").run {
            init(ECGenParameterSpec("secp256r1"))
            getParameterSpec(ECParameterSpec::class.java)
        }
        return order == expected.order && cofactor == expected.cofactor &&
            generator == expected.generator && curve == expected.curve
    }

    private fun isLowercaseHex(value: String, length: Int): Boolean =
        value.length == length && value.all { it in '0'..'9' || it in 'a'..'f' }

    private const val CLIENT_CONTEXT = "AetherLink initial pairing client proof v1"
    private const val RESULT_CONTEXT = "AetherLink initial pairing runtime result proof v1"
    private val P256_ORDER = BigInteger("ffffffff00000000ffffffffffffffffbce6faada7179e84f3b9cac2fc632551", 16)
}

File: apps/macos/Pairing/Sources/InitialPairingProof.swift
import CryptoKit
import Foundation

public enum InitialPairingProofError: Error, Equatable, LocalizedError, Sendable {
    case invalidField(String)
    case invalidPublicKey(String)
    case invalidFingerprint(String)
    case invalidSignature
    case rejectedResultCannotBeSigned

    public var errorDescription: String? {
        switch self {
        case .invalidField(let field):
            return "The initial pairing proof field \(field) is invalid."
        case .invalidPublicKey(let field):
            return "The initial pairing proof public key \(field) is invalid."
        case .invalidFingerprint(let field):
            return "The initial pairing proof fingerprint \(field) is invalid."
        case .invalidSignature:
            return "The initial pairing proof signature is invalid."
        case .rejectedResultCannotBeSigned:
            return "Only accepted initial pairing results may be signed."
        }
    }
}

public protocol InitialPairingRuntimeResultSigning: Sendable {
    func signInitialPairingResult(
        _ result: InitialPairingRuntimeResult
    ) throws -> InitialPairingRuntimeResultProof
}

public enum InitialPairingProof {
    public static let scheme = "p256-sha256-der-v1"
    public static let protocolVersion = 1
    public static let clientContext = "AetherLink initial pairing client proof v1"
    public static let runtimeResultContext = "AetherLink initial pairing runtime result proof v1"

    public static func isCanonicalTransportBinding(_ value: String) -> Bool {
        value == "none" || isLowercaseHex(value, count: 64)
    }

    public static func isCanonicalDigest(_ value: String) -> Bool {
        isLowercaseHex(value, count: 64)
    }

    static func messageData(context: String, fields: [(String, String)]) -> Data {
        let components = [context] + fields.flatMap { name, value in
            [name, String(value.utf8.count), value]
        }
        return Data(components.joined(separator: "\n").utf8)
    }

    static func digestHex(_ data: Data) -> String {
        SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
    }

    static func validatedPublicKey(
        base64: String,
        fingerprint: String,
        keyField: String,
        fingerprintField: String
    ) throws -> P256.Signing.PublicKey {
        guard let data = canonicalBase64Data(base64),
              let key = try? P256.Signing.PublicKey(derRepresentation: data),
              key.derRepresentation == data else {
            throw InitialPairingProofError.invalidPublicKey(keyField)
        }
        guard isCanonicalDigest(fingerprint), digestHex(data) == fingerprint else {
            throw InitialPairingProofError.invalidFingerprint(fingerprintField)
        }
        return key
    }

    static func validatedSignature(_ base64: String) throws -> P256.Signing.ECDSASignature {
        guard let data = canonicalBase64Data(base64),
              let signature = try? P256.Signing.ECDSASignature(derRepresentation: data),
              signature.derRepresentation == data else {
            throw InitialPairingProofError.invalidSignature
        }
        return signature
    }

    private static func canonicalBase64Data(_ value: String) -> Data? {
        guard let data = Data(base64Encoded: value), data.base64EncodedString() == value else {
            return nil
        }
        return data
    }

    private static func isLowercaseHex(_ value: String, count: Int) -> Bool {
        value.utf8.count == count && value.utf8.allSatisfy {
            (UInt8(ascii: "0")...UInt8(ascii: "9")).contains($0) ||
                (UInt8(ascii: "a")...UInt8(ascii: "f")).contains($0)
        }
    }
}

public struct InitialPairingClientProof: Equatable, Sendable {
    public var scheme: String
    public var protocolVersion: Int
    public var requestID: String
    public var pairingNonce: String
    public var pairingCode: String
    public var runtimeDeviceID: String
    public var runtimePublicKey: String
    public var runtimeKeyFingerprint: String
    public var clientDeviceID: String
    public var clientDeviceName: String
    public var clientPublicKey: String
    public var clientKeyFingerprint: String
    public var transportBinding: String
    public var signatureBase64: String

    public init(
        scheme: String = InitialPairingProof.scheme,
        protocolVersion: Int = InitialPairingProof.protocolVersion,
        requestID: String,
        pairingNonce: String,
        pairingCode: String,
        runtimeDeviceID: String,
        runtimePublicKey: String,
        runtimeKeyFingerprint: String,
        clientDeviceID: String,
        clientDeviceName: String,
        clientPublicKey: String,
        clientKeyFingerprint: String,
        transportBinding: String,
        signatureBase64: String
    ) throws {
        self.scheme = scheme
        self.protocolVersion = protocolVersion
        self.requestID = requestID
        self.pairingNonce = pairingNonce
        self.pairingCode = pairingCode
        self.runtimeDeviceID = runtimeDeviceID
        self.runtimePublicKey = runtimePublicKey
        self.runtimeKeyFingerprint = runtimeKeyFingerprint
        self.clientDeviceID = clientDeviceID
        self.clientDeviceName = clientDeviceName
        self.clientPublicKey = clientPublicKey
        self.clientKeyFingerprint = clientKeyFingerprint
        self.transportBinding = transportBinding
        self.signatureBase64 = signatureBase64
        try validateShape()
    }

    public func signedMessageData() -> Data {
        InitialPairingProof.messageData(context: InitialPairingProof.clientContext, fields: fields)
    }

    public static func signingMessageData(
        scheme: String = InitialPairingProof.scheme,
        protocolVersion: Int = InitialPairingProof.protocolVersion,
        requestID: String,
        pairingNonce: String,
        pairingCode: String,
        runtimeDeviceID: String,
        runtimePublicKey: String,
        runtimeKeyFingerprint: String,
        clientDeviceID: String,
        clientDeviceName: String,
        clientPublicKey: String,
        clientKeyFingerprint: String,
        transportBinding: String
    ) throws -> Data {
        guard scheme == InitialPairingProof.scheme else {
            throw InitialPairingProofError.invalidField("scheme")
        }
        guard protocolVersion == InitialPairingProof.protocolVersion else {
            throw InitialPairingProofError.invalidField("protocol_version")
        }
        guard InitialPairingProof.isCanonicalTransportBinding(transportBinding) else {
            throw InitialPairingProofError.invalidField("transport_binding")
        }
        _ = try InitialPairingProof.validatedPublicKey(
            base64: runtimePublicKey,
            fingerprint: runtimeKeyFingerprint,
            keyField: "runtime_public_key",
            fingerprintField: "runtime_key_fingerprint"
        )
        _ = try InitialPairingProof.validatedPublicKey(
            base64: clientPublicKey,
            fingerprint: clientKeyFingerprint,
            keyField: "client_public_key",
            fingerprintField: "client_key_fingerprint"
        )
        return InitialPairingProof.messageData(
            context: InitialPairingProof.clientContext,
            fields: [
                ("scheme", scheme), ("protocol_version", String(protocolVersion)),
                ("request_id", requestID), ("pairing_nonce", pairingNonce),
                ("pairing_code", pairingCode), ("runtime_device_id", runtimeDeviceID),
                ("runtime_public_key", runtimePublicKey),
                ("runtime_key_fingerprint", runtimeKeyFingerprint),
                ("client_device_id", clientDeviceID), ("client_device_name", clientDeviceName),
                ("client_public_key", clientPublicKey),
                ("client_key_fingerprint", clientKeyFingerprint),
                ("transport_binding", transportBinding),
            ]
        )
    }

    public func requestDigest() -> String {
        InitialPairingProof.digestHex(signedMessageData())
    }

    public func verify() -> Bool {
        do {
            try validateShape()
            let key = try InitialPairingProof.validatedPublicKey(
                base64: clientPublicKey,
                fingerprint: clientKeyFingerprint,
                keyField: "client_public_key",
                fingerprintField: "client_key_fingerprint"
            )
            let signature = try InitialPairingProof.validatedSignature(signatureBase64)
            return key.isValidSignature(signature, for: SHA256.hash(data: signedMessageData()))
        } catch {
            return false
        }
    }

    private var fields: [(String, String)] {
        [
            ("scheme", scheme), ("protocol_version", String(protocolVersion)),
            ("request_id", requestID), ("pairing_nonce", pairingNonce),
            ("pairing_code", pairingCode), ("runtime_device_id", runtimeDeviceID),
            ("runtime_public_key", runtimePublicKey), ("runtime_key_fingerprint", runtimeKeyFingerprint),
            ("client_device_id", clientDeviceID), ("client_device_name", clientDeviceName),
            ("client_public_key", clientPublicKey), ("client_key_fingerprint", clientKeyFingerprint),
            ("transport_binding", transportBinding),
        ]
    }

    private func validateShape() throws {
        guard scheme == InitialPairingProof.scheme else { throw InitialPairingProofError.invalidField("scheme") }
        guard protocolVersion == InitialPairingProof.protocolVersion else { throw InitialPairingProofError.invalidField("protocol_version") }
        guard InitialPairingProof.isCanonicalTransportBinding(transportBinding) else { throw InitialPairingProofError.invalidField("transport_binding") }
        _ = try InitialPairingProof.validatedPublicKey(base64: runtimePublicKey, fingerprint: runtimeKeyFingerprint, keyField: "runtime_public_key", fingerprintField: "runtime_key_fingerprint")
        _ = try InitialPairingProof.validatedPublicKey(base64: clientPublicKey, fingerprint: clientKeyFingerprint, keyField: "client_public_key", fingerprintField: "client_key_fingerprint")
        _ = try InitialPairingProof.validatedSignature(signatureBase64)
    }
}

public struct InitialPairingRuntimeResult: Equatable, Sendable {
    public var scheme: String
    public var protocolVersion: Int
    public var requestID: String
    public var pairingRequestDigest: String
    public var accepted: Bool
    public var runtimeDeviceID: String
    public var runtimePublicKey: String
    public var runtimeKeyFingerprint: String
    public var trustedDeviceID: String
    public var message: String
    public var transportBinding: String

    public init(
        scheme: String = InitialPairingProof.scheme,
        protocolVersion: Int = InitialPairingProof.protocolVersion,
        requestID: String,
        pairingRequestDigest: String,
        accepted: Bool,
        runtimeDeviceID: String,
        runtimePublicKey: String,
        runtimeKeyFingerprint: String,
        trustedDeviceID: String,
        message: String,
        transportBinding: String
    ) throws {
        guard accepted else { throw InitialPairingProofError.rejectedResultCannotBeSigned }
        guard scheme == InitialPairingProof.scheme else { throw InitialPairingProofError.invalidField("scheme") }
        guard protocolVersion == InitialPairingProof.protocolVersion else { throw InitialPairingProofError.invalidField("protocol_version") }
        guard InitialPairingProof.isCanonicalDigest(pairingRequestDigest) else { throw InitialPairingProofError.invalidField("pairing_request_digest") }
        guard InitialPairingProof.isCanonicalTransportBinding(transportBinding) else { throw InitialPairingProofError.invalidField("transport_binding") }
        _ = try InitialPairingProof.validatedPublicKey(base64: runtimePublicKey, fingerprint: runtimeKeyFingerprint, keyField: "runtime_public_key", fingerprintField: "runtime_key_fingerprint")
        self.scheme = scheme
        self.protocolVersion = protocolVersion
        self.requestID = requestID
        self.pairingRequestDigest = pairingRequestDigest
        self.accepted = accepted
        self.runtimeDeviceID = runtimeDeviceID
        self.runtimePublicKey = runtimePublicKey
        self.runtimeKeyFingerprint = runtimeKeyFingerprint
        self.trustedDeviceID = trustedDeviceID
        self.message = message
        self.transportBinding = transportBinding
    }

    public func signedMessageData() -> Data {
        InitialPairingProof.messageData(context: InitialPairingProof.runtimeResultContext, fields: [
            ("scheme", scheme), ("protocol_version", String(protocolVersion)),
            ("request_id", requestID), ("pairing_request_digest", pairingRequestDigest),
            ("accepted", accepted ? "true" : "false"), ("runtime_device_id", runtimeDeviceID),
            ("runtime_public_key", runtimePublicKey), ("runtime_key_fingerprint", runtimeKeyFingerprint),
            ("trusted_device_id", trustedDeviceID), ("message", message),
            ("transport_binding", transportBinding),
        ])
    }

    public func resultDigest() -> String {
        InitialPairingProof.digestHex(signedMessageData())
    }

    func validate() throws {
        guard accepted else { throw InitialPairingProofError.rejectedResultCannotBeSigned }
        guard scheme == InitialPairingProof.scheme else { throw InitialPairingProofError.invalidField("scheme") }
        guard protocolVersion == InitialPairingProof.protocolVersion else { throw InitialPairingProofError.invalidField("protocol_version") }
        guard InitialPairingProof.isCanonicalDigest(pairingRequestDigest) else { throw InitialPairingProofError.invalidField("pairing_request_digest") }
        guard InitialPairingProof.isCanonicalTransportBinding(transportBinding) else { throw InitialPairingProofError.invalidField("transport_binding") }
        _ = try InitialPairingProof.validatedPublicKey(base64: runtimePublicKey, fingerprint: runtimeKeyFingerprint, keyField: "runtime_public_key", fingerprintField: "runtime_key_fingerprint")
    }
}

public struct InitialPairingRuntimeResultProof: Equatable, Sendable {
    public var result: InitialPairingRuntimeResult
    public var signatureBase64: String

    public init(result: InitialPairingRuntimeResult, signatureBase64: String) throws {
        _ = try InitialPairingProof.validatedSignature(signatureBase64)
        self.result = result
        self.signatureBase64 = signatureBase64
    }

    public func verify() -> Bool {
        do {
            try result.validate()
            let key = try InitialPairingProof.validatedPublicKey(
                base64: result.runtimePublicKey,
                fingerprint: result.runtimeKeyFingerprint,
                keyField: "runtime_public_key",
                fingerprintField: "runtime_key_fingerprint"
            )
            let signature = try InitialPairingProof.validatedSignature(signatureBase64)
            return key.isValidSignature(signature, for: SHA256.hash(data: result.signedMessageData()))
        } catch {
            return false
        }
    }
}
